pennylane/clusters.py

```diff
@@ -0,0 +1,476 @@
+        ):  # pylint: disable=dangerous-default-value, function-redefined
+            template(params, wires=_wires, **kwargs)
+            return MEASURE_MAP[_m](_obs)
+
+        qnode = QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        qnodes.append(qnode)
+
+    return qnodes
+
+
+def apply(func, qnode_cluster):
+    """Lazily apply a function to the constituent QNodes of a :class:`QNodeCluster`.
+
+    Args:
+        func (callable): A function to be applied to the QNodeCluster results.
+            This function must be supported by the corresponding QNodeCluster
+            interface; i.e., a ``torch`` QNodeCluster can only be acted on functions
+            that accept ``torch.tensor`` objects.
+        qnode_cluster (QNodeCluster): a QNode cluster.
+
+    **Example:**
+
+    We can create a QNodeCluster using :func:`~.map`:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)]
+    >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, dev, interface="torch")
+
+    As we are using the ``'torch'`` interface, we now apply ``torch.sum``
+    to the QNodeCluster:
+
+    >>> cost = qml.apply(torch.sum, qnodes)
+
+    This is a lazy composition --- no QNode evaluation has yet occured. Evaluation
+    only occurs when the returned function ``cost`` is evaluated:
+
+    >>> x = qml.init.strong_ent_layers_normal(3, 2)
+    >>> cost(x)
+    tensor(0.9092, dtype=torch.float64, grad_fn=<SumBackward0>)
+    """
+    return lambda params, **kwargs: func(qnode_cluster(params, **kwargs))
+
+
+def sum(x):
+    """Lazily sum the constituent QNodes of a :class:`QNodeCluster`.
+
+    Args:
+        x (QNodeCluster): a QNode cluster of independent QNodes.
+
+    .. seealso:: :func:`~.apply`, :func:`~.dot`
+
+    **Example:**
+
+    We can create a QNodeCluster using :func:`~.map`:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)]
+    >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, dev, interface="torch")
+
+    For the cost function, we now sum the results of all QNodes in the cluster:
+
+    >>> cost = qml.sum(qnodes)
+
+    This is a lazy summation --- no QNode evaluation has yet occured. Evaluation
+    only occurs when the returned function ``cost`` is evaluated:
+
+    >>> x = qml.init.strong_ent_layers_normal(3, 2)
+    >>> cost(x)
+    tensor(0.9092, dtype=torch.float64, grad_fn=<SumBackward0>)
+    """
+    if hasattr(x, "interface") and x.interface is not None:
+        if x.interface == "tf":
+            import tensorflow as tf
+
+            return lambda params, **kwargs: tf.reduce_sum(x(params, **kwargs))
+
+        if x.interface == "torch":
+            import torch
+
+            return lambda params, **kwargs: torch.sum(x(params, **kwargs))
+
+        if x.interface in ("autograd", "numpy"):
```

Comment:
Currently, though it supports backward compatibility, the ``self.interface == "numpy"`` might be mistaken for ``self.interface==None``

Reply:
Yep, this is the main reason we are slowly deprecating `numpy` and moving to `autograd`. It can't be removed fully until all the tutorials and docs are updated, though (so it's still a todo). At some point, we should modify the `qml.qnodes.QNode` function so that using `numpy` raises a deprecation warning.

Reply:
That makes sense! Perhaps adding a ``TODO`` comment could be in place, such that even when deprecated, this part of the code would not be forgotten

Reply:
Done!